server/version.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/juju/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// VersionFeature is a unique identifier for minimum support version.
+type VersionFeature int
+
+// Version Fetures.
+const (
+	VersionBase VersionFeature = iota
+	VersionRegionMergeAndRaftLearner
+	VersionBatchSplit
+)
+
+var featuresDict = map[VersionFeature]Version{
+	VersionBase:                      {Major: 1},
+	VersionRegionMergeAndRaftLearner: {Major: 2, Minor: 0},
+	VersionBatchSplit:                {Major: 2, Minor: 1},
+}
+
+// TargetVersion get target version by version feature
+func TargetVersion(v VersionFeature) Version {
+	target, ok := featuresDict[v]
+	if !ok {
+		log.Fatalf("version not exist, feature %d", v)
+	}
+	return target
+}
+
+// Version record version information.
+type Version struct {
+	Major         int32
+	Minor         int32
+	Patch         int32
+	Unstable      int32
+	UnstablePatch int32
+}
+
+// Less compare two version.
+func (v *Version) Less(ov Version) bool {
+	if v.Major < ov.Major {
+		return true
+	} else if v.Major > ov.Major {
+		return false
+	}
+	if v.Minor < ov.Minor {
+		return true
+	} else if v.Minor > ov.Minor {
+		return false
+	}
+	if v.Patch < ov.Patch {
+		return true
+	} else if v.Patch > ov.Patch {
+		return false
+	}
+	if v.Unstable < ov.Unstable {
+		return true
+	} else if v.Unstable > ov.Unstable {
+		return false
+	}
+	if v.UnstablePatch < ov.UnstablePatch {
+		return true
+	}
+	return false
+}
+
+func (v Version) String() string {
+	convMap := map[int32]string{
+		1: "alpha",
+		2: "beta",
+		3: "rc",
```

Comment:
In fact the pre-release version can be any string contains `[0-9A-Za-z-]`. See https://semver.org/#spec-item-9